### PR TITLE
Address #138 & #139

### PR DESCRIPTION
**Pull request by xiaomi7732**

### CODE REVIEW

**Comment by xiaomi7732** [comment]
This PR is supposed to addresses #138 & #139. Please review! Thanks!

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipelineFactory.cs [hunk 1]
```diff
@@ -40,7 +40,13 @@ public static DiagnosticPipeline CreatePipeline(IConfiguration configuration, IH
                 healthReporter = CreateHealthReporter(configuration);
             }
             Verify.Operation(healthReporter != null, $"An instance of a health reporter could not be created and none was provider as a parameter to {nameof(CreatePipeline)} method");
-            (healthReporter as IRequireActivation)?.Activate();
+            bool? activateResult = (healthReporter as IRequireActivation)?.Activate();
```
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/DiagnosticPipelineFactory.cs]: Let's not change the contract if not absolutely necessary.

How about throwing an exception upon activation failure? An exception can happen anyway, depending on health reporter implementation, so there is no loss of generality here.

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/HealthReporters/CsvHealthReporter.cs [hunk 2]
```diff
@@ -111,60 +119,79 @@ public CsvHealthReporter(string configurationFilePath)
...
         /// because CreateStreamWriter calls into virtual mehtod of GetReportFileName().
         /// </remarks>
-        public void Activate()
+        public bool Activate()
         {
             VerifyObjectIsNotDisposed();
             SetNewStreamWriter();
+            if (StreamWriter == null)
+            {
+                return false;
```
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/HealthReporters/CsvHealthReporter.cs]: throw new InvalidOperationException (or something like that)

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/HealthReporters/CsvHealthReporter.cs [hunk 3]
```diff
@@ -290,9 +372,18 @@ private void Initialize(CsvHealthReporterConfiguration configuration, INewReport
...
             this.Configuration = configuration;
 
-            // Create a default throttle
+            // Create a default throttle.
             this.throttle = new TimeSpanThrottle(TimeSpan.FromMilliseconds(DefaultThrottlingPeriodMsec));
 
+            // Set the file size for csv health report. Minimum is 1MB. Default is 8192MB.
+            this.SingleLogFileMaximumSizeInBytes = (long)(configuration.SingleLogFileMaximumSizeInMBytes > 0 ? configuration.SingleLogFileMaximumSizeInMBytes : 8192) * 1024 * 1024;
+
+            // Set default value for retention days for the logs. Minimum value is 1.
```
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Core/Implementations/HealthReporters/CsvHealthReporter.cs]: Minimum value is 30

File: src/Microsoft.Diagnostics.EventFlow.Inputs.Etw/EtwInput.cs [hunk 4]
```diff
@@ -76,7 +76,7 @@ public void Activate()
             if (this.Providers.Count() == 0)
             {
                 healthReporter.ReportWarning($"{nameof(EtwInput)}: no providers configured", nameof(EtwInput));
-                return;
+                return false;
```
Comment (karolz-ms) [inline src/Microsoft.Diagnostics.EventFlow.Inputs.Etw/EtwInput.cs]: revert please

**Comment by xiaomi7732** [comment]
Had an offline discussion with @karolz-ms. We agreed on the following bases:
* The CsvHealthReporter does not throw when it cannot create a file—but if it can’t, it will not report anything (it becomes IdleHealthReporter, effectively)
* Have a “debugMode” setting for CsvHealthReporter (off by default)
* If debugMode is on, throw exception from Activate if the reporter cannot create a file for writing
All these changes shall be implemented within the CsvHealthReporter, with no impact on existing interfaces or the existing pipeline factories.

I'll update the implementation based on that, and of course, the comments from the code review.

File: README.md [hunk 5]
```diff
@@ -848,7 +848,9 @@ This health reporter writes all errors, warnings, and informational traces gener
     "logFileFolder": ".",
     "logFilePrefix": "HealthReport",
     "minReportLevel": "Warning",
-    "throttlingPeriodMsec": "1000"
+    "throttlingPeriodMsec": "1000",
+    "singleLogFileMaximumSizeInMBytes": "8192",
+    "retentionLogsInDays": "30"
```
[the remark this answers is not in the document]
Comment (xiaomi7732) [reply README.md]: Makes sens. Fixed in i2.

File: README.md [hunk 6]
```diff
@@ -858,8 +860,11 @@ This health reporter writes all errors, warnings, and informational traces gener
...
 | `throttlingPeriodMsec` | number of milliseconds | No | Specifies the throttling time period. This setting protects the health reporter from being overwhelmed, which can happen if a message is repeatedly generated due to an error in the pipeline. Default is 0, for no throttling. |
+| `singleLogFileMaximumSizeInMBytes` | File size in MB/number | No | Specifies the size of the log file in MB before rotating happens. The default value is 8192 MB (8 GB). Once the size of log file exceeds the value, it will be renamed from fileName.csv to fileName_last.csv. Then logs will be written to a new fileName.csv. This setting prevents a single log file become too big. |
+| `retentionLogsInDays` | number of days for the logs files retain | No | Specifies the days for the log files to be retained. The default value is 30. Any log file created before the number of days will be removed automatically. This prevents continuous generating logs taking up the storage. |
```
[the remark this answers is not in the document]
Comment (xiaomi7732) [reply README.md]: Thanks! Fixed in the next iteration.

File: README.md [hunk 7]
```diff
@@ -858,8 +860,11 @@ This health reporter writes all errors, warnings, and informational traces gener
...
 
-CsvHealthReporter will try to open the log file for writing during initialization. If it can't, it is considered a fatal error and will cause the whole EventFlow pipeline to fail. This can happen especially if a value for the log file path is not provided (default is used, which is application executables folder) and the application executables are residing on a read-only file system. Docker tools for Visual Studio use this configuration during debugging, so for containerized services the recommended practice is to specify the log file path explicitly.
+
+CsvHealthReporter will try to open the log file for writing during initialization. If it can't, a debug message will be output to the debugger viewer like Visual Studio Output window, etc. Docker tools for Visual Studio use this configuration during debugging, so for containerized services the recommended practice is to specify the log file path explicitly.
```
[the remark this answers is not in the document]
Comment (xiaomi7732) [reply README.md]: Yup. Fixed in next iteration.

File: src/Microsoft.Diagnostics.EventFlow.Core/Implementations/HealthReporters/CsvHealthReporter.cs (hunk 3 above)
Comment (xiaomi7732) [reply src/Microsoft.Diagnostics.EventFlow.Core/Implementations/HealthReporters/CsvHealthReporter.cs]: @karolz-ms , do you really mean to make the minimum value to 30? Which sounds a bit aggressive. I think a minimum of 1 day but default to 30 days is more flexible.

Comment (xiaomi7732) [reply src/Microsoft.Diagnostics.EventFlow.Core/Implementations/HealthReporters/CsvHealthReporter.cs]: @karolz-ms , do you really mean to make the minimum value to 30? Which sounds a bit aggressive. I think a minimum of 1 day but default to 30 days is more flexible.

**Comment by xiaomi7732** [comment]
@karolz-ms, I believe I addressed all the issues. Please let me know if you see anything outstanding.
Thanks!

**Pull request merged**